core/primitives/src/types.rs

```diff
@@ -81,37 +81,68 @@ pub struct ViewCallResult {
     pub result: Vec<u8>,
 }
 
-/// TODO: Call non-view function in the contracts.
 #[derive(Hash, Serialize, Deserialize, PartialEq, Eq, Debug, Clone)]
-pub struct TransactionBody {
+pub struct StakeTransaction {
+    pub nonce: u64,
+    pub staker: AccountId,
+    pub amount: u64,
+}
+
+#[derive(Hash, Serialize, Deserialize, PartialEq, Eq, Debug, Clone)]
+pub struct SendMoneyTransaction {
     pub nonce: u64,
     pub sender: AccountId,
     pub receiver: AccountId,
     pub amount: u64,
+}
+
+#[derive(Hash, Serialize, Deserialize, PartialEq, Eq, Debug, Clone)]
+pub struct DeployContractTransaction {
+    pub nonce: u64,
+    pub owner: AccountId,
+    pub contract_id: AccountId,
+    pub wasm_byte_array: Vec<u8>,
+}
+
+#[derive(Hash, Serialize, Deserialize, PartialEq, Eq, Debug, Clone)]
+pub struct FunctionCallTransaction {
```

Comment:
function call needs to have amount as well

Reply:
Also it might be better to have sender and receiver outside transaction body, as every type of transaction needs to have sender and receiver.

Reply:
amount is currently unused. https://github.com/nearprotocol/nearcore/pull/151/files#diff-5e5e1c3aec9ddfde0a9054d062ab3db9L360

sender is required for every type but means different things in different contexts. receiver is not required for every type and is currently only relevant for `SendMoney`. in other contexts it is:
```
CallFunction => contract_name
DeployContract => contract_name
Stake => n/a
```

Reply:
amount will be used soon. For receiver, I tend to think of the contract that is called as the receiver and the contract who invokes the call as the sender. For staking, the receiver will be the same as the sender. But I guess it's also fine as it is here.

Reply:
what will amount be used for in the function call context? the deploy contract context?

for staking, having the receiver is redundant.

in general, I don't think we are losing anything by making things more consice for specific use cases.